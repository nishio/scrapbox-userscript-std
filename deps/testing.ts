export * from "https://deno.land/std@0.217.0/testing/asserts.ts";
export * from "https://deno.land/std@0.217.0/testing/snapshot.ts";
