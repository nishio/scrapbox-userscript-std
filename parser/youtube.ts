// ported from https://github.com/takker99/ScrapBubble/blob/0.4.0/Page.tsx#L662

/** Regular expressions for matching different YouTube URL formats */
// Matches standard youtube.com/watch URLs (including music.youtube.com)
const youtubeRegExp = /https?:\/\/(?:www\.|music\.|)youtube\.com\/watch/;
// Matches short youtu.be URLs with optional query parameters
const youtubeDotBeRegExp =
  /https?:\/\/youtu\.be\/([a-zA-Z\d_-]+)(?:\?([^\s]{0,100})|)/;
// Matches YouTube Shorts URLs
const youtubeShortRegExp =
  /https?:\/\/(?:www\.|)youtube\.com\/shorts\/([a-zA-Z\d_-]+)(?:\?([^\s]+)|)/;
// Matches playlist URLs (including music.youtube.com playlists)
const youtubeListRegExp =
  /https?:\/\/(?:www\.|music\.|)youtube\.com\/playlist\?((?:[^\s]+&|)list=([a-zA-Z\d_-]+)(?:&[^\s]+|))/;

/** Properties extracted from a YouTube URL
 * This type represents the parsed data from different types of YouTube URLs.
 * It's a union type that handles both video-related URLs and playlist URLs.
 *
 * For video URLs (standard, short URLs, or youtu.be links):
 * @property params - URL query parameters (e.g., timestamp, playlist reference)
 * @property videoId - The unique identifier of the video
 * @property pathType - The URL format type:
 *   - "com": Standard youtube.com/watch?v= format
 *   - "dotbe": Short youtu.be/ format
 *   - "short": YouTube Shorts format
 *
 * For playlist URLs:
 * @property params - URL query parameters
 * @property listId - The unique identifier of the playlist
 * @property pathType - Always "list" for playlist URLs
 */
export type YoutubeProps = {
  params: URLSearchParams;
  videoId: string;
  pathType: "com" | "dotbe" | "short";
} | {
  params: URLSearchParams;
  listId: string;
  pathType: "list";
};

/** Parse a YouTube URL to extract video/playlist ID and other properties
 *
 * This function handles various YouTube URL formats:
 * 1. Standard video URLs:
 *    - https://www.youtube.com/watch?v={videoId}
 *    - https://music.youtube.com/watch?v={videoId}
 *
 * 2. Short URLs:
 *    - https://youtu.be/{videoId}
 *    - Can include optional query parameters
 *
 * 3. YouTube Shorts:
 *    - https://youtube.com/shorts/{videoId}
 *    - https://www.youtube.com/shorts/{videoId}
 *
 * 4. Playlist URLs:
 *    - https://youtube.com/playlist?list={listId}
 *    - https://music.youtube.com/playlist?list={listId}
 *
 * The function preserves all query parameters from the original URL.
 *
 * @param url - Any URL or string to parse
 * @returns A {@linkcode Result}<{@linkcode YoutubeProps}, {@linkcode undefined}> containing:
 *          - Success: The extracted video/playlist information with:
 *            - For videos: videoId, params, and pathType ("com", "dotbe", or "short")
 *            - For playlists: listId, params, and pathType ("list")
 *          - Error: {@linkcode undefined} if not a valid YouTube URL
 */
export const parseYoutube = (url: string): YoutubeProps | undefined => {
  if (youtubeRegExp.test(url)) {
    const params = new URL(url).searchParams;
    const videoId = params.get("v");
    if (videoId) {
      return {
        pathType: "com",
        videoId,
        params,
      };
    }
  }

  {
    const matches = url.match(youtubeDotBeRegExp);
    if (matches) {
      const [, videoId, params] = matches;
      return {
        videoId,
        params: new URLSearchParams(params),
        pathType: "dotbe",
      };
    }
  }

  {
    const matches = url.match(youtubeShortRegExp);
    if (matches) {
      const [, videoId, params] = matches;
      return {
        videoId,
        params: new URLSearchParams(params),
        pathType: "short",
      };
    }
  }

  {
    const matches = url.match(youtubeListRegExp);
    if (matches) {
      const [, params, listId] = matches;

      return { listId, params: new URLSearchParams(params), pathType: "list" };
    }
  }

  return undefined;
};
