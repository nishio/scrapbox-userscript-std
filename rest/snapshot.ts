import type {
  ErrorLike,
  NotFoundError,
  NotLoggedInError,
  NotMemberError,
  PageSnapshotList,
  PageSnapshotResult,
} from "@cosense/types/rest";
import { cookie } from "./auth.ts";
import { type BaseOptions, setDefaults } from "./options.ts";
import { parseHTTPError } from "./parseHTTPError.ts";
import {
  isErr,
  mapAsyncForResult,
  mapErrAsyncForResult,
  type Result,
  unwrapOk,
} from "option-t/plain_result";
import { type HTTPError, responseIntoResult } from "./responseIntoResult.ts";
import type { FetchError } from "./mod.ts";

/** Error that occurs when an invalid `timestampId` is provided to {@linkcode getSnapshot} */
export interface InvalidPageSnapshotIdError extends ErrorLike {
  name: "InvalidPageSnapshotIdError";
}

export type SnapshotError =
  | NotFoundError
  | NotLoggedInError
  | NotMemberError
  | InvalidPageSnapshotIdError
  | HTTPError;

/** get a page snapshot
 *
 * @param options `connect.sid` etc.
 */
export const getSnapshot = async (
  project: string,
  pageId: string,
  timestampId: string,
  options?: BaseOptions,
): Promise<Result<PageSnapshotResult, SnapshotError | FetchError>> => {
  const { sid, hostName, fetch } = setDefaults(options ?? {});

  const req = new Request(
    `https://${hostName}/api/page-snapshots/${project}/${pageId}/${timestampId}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );

  const res = await fetch(req);
  if (isErr(res)) return res;

  return mapAsyncForResult(
    await mapErrAsyncForResult(
      responseIntoResult(unwrapOk(res)),
      async (error) =>
        error.response.status === 422
          ? {
            name: "InvalidPageSnapshotIdError",
            message: await error.response.text(),
          }
          : (await parseHTTPError(error, [
            "NotFoundError",
            "NotLoggedInError",
            "NotMemberError",
          ])) ?? error,
    ),
    (res) => res.json() as Promise<PageSnapshotResult>,
  );
};

export type SnapshotTimestampIdsError =
  | NotFoundError
  | NotLoggedInError
  | NotMemberError
  | HTTPError;

/**
 * Retrieves the timestamp IDs for a specific page in a project.
 *
 * @param project - The name of the project.
 * @param pageId - The ID of the page.
 * @param options - Optional configuration options.
 * @returns A promise that resolves to a {@linkcode Result} object containing the page snapshot list if successful,
 * or an error if the request fails.
 */
export const getTimestampIds = async (
  project: string,
  pageId: string,
  options?: BaseOptions,
): Promise<
  Result<PageSnapshotList, SnapshotTimestampIdsError | FetchError>
> => {
  const { sid, hostName, fetch } = setDefaults(options ?? {});

  const req = new Request(
    `https://${hostName}/api/page-snapshots/${project}/${pageId}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );

  const res = await fetch(req);
  if (isErr(res)) return res;

  return mapAsyncForResult(
    await mapErrAsyncForResult(
      responseIntoResult(unwrapOk(res)),
      async (error) =>
        (await parseHTTPError(error, [
          "NotFoundError",
          "NotLoggedInError",
          "NotMemberError",
        ])) ?? error,
    ),
    (res) => res.json() as Promise<PageSnapshotList>,
  );
};
