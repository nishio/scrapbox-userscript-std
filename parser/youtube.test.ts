import { parseYoutube } from "./youtube.ts";
import { assertSnapshot } from "@std/testing/snapshot";

/** Test suite for YouTube URL parsing functionality
 * This test suite verifies the parseYoutube function's ability to handle various
 * YouTube URL formats and invalid inputs using snapshot testing.
 */
Deno.test({
  name: "youtube links",
  ignore: true,
  fn: async (t) => {
  /** Test valid YouTube URL formats
   * Verifies parsing of:
   * - Standard watch URLs (youtube.com/watch?v=...)
   * - Playlist URLs (youtube.com/playlist?list=...)
   * - Watch URLs within playlists
   * - YouTube Music URLs (music.youtube.com)
   * - Short URLs (youtu.be/...)
   */
  await t.step("is", async (t) => {
    await assertSnapshot(
      t,
      parseYoutube("https://www.youtube.com/watch?v=LSvaOcaUQ3Y"),
    );
    await assertSnapshot(
      t,
      parseYoutube(
        "https://www.youtube.com/playlist?list=PLmoRDY8IgE2Okxy4WWdP95RHXOTGzJfQs",
      ),
    );
    await assertSnapshot(
      t,
      parseYoutube(
        "https://www.youtube.com/watch?v=57rdbK4vmKE&list=PLmoRDY8IgE2Okxy4WWdP95RHXOTGzJfQs",
      ),
    );
    await assertSnapshot(
      t,
      parseYoutube(
        "https://music.youtube.com/watch?v=nj1cre2e6t0",
      ),
    );
    await assertSnapshot(
      t,
      parseYoutube(
        "https://youtu.be/nj1cre2e6t0",
      ),
    );
  });

  /** Test invalid URL formats
   * Verifies that the function correctly returns undefined for:
   * - URLs from other services (e.g., Gyazo)
   * - Non-URL strings (including Japanese text)
   * - Similar but invalid domains (e.g., "yourtube.com")
   * - Generic URLs
   */
  await t.step("is not", async (t) => {
    await assertSnapshot(
      t,
      parseYoutube(
        "https://gyazo.com/da78df293f9e83a74b5402411e2f2e01",
      ),
    );
    await assertSnapshot(
      t,
      parseYoutube(
        "test_text",
      ),
    );
    await assertSnapshot(
      t,
      parseYoutube(
        "https://yourtube.com/watch?v=rafere",
      ),
    );
    await assertSnapshot(
      t,
      parseYoutube(
        "https://example.com",
      ),
    );
  });
});
