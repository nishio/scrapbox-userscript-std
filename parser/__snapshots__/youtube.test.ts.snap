export const snapshot = {};

snapshot[`youtube links > is 1`] = `
{
  params: URLSearchParams {
    [Symbol("[[webidl.brand]]")]: Symbol("[[webidl.brand]]"),
    [Symbol("url object")]: URL {
      hash: "",
      host: "www.youtube.com",
      hostname: "www.youtube.com",
      href: "https://www.youtube.com/watch?v=LSvaOcaUQ3Y",
      origin: "https://www.youtube.com",
      password: "",
      pathname: "/watch",
      port: "",
      protocol: "https:",
      search: "?v=LSvaOcaUQ3Y",
      username: "",
    },
    [Symbol(list)]: [
      [
        "v",
        "LSvaOcaUQ3Y",
      ],
    ],
  },
  pathType: "com",
  videoId: "LSvaOcaUQ3Y",
}
`;

snapshot[`youtube links > is 2`] = `
{
  listId: "PLmoRDY8IgE2Okxy4WWdP95RHXOTGzJfQs",
  params: URLSearchParams {
    [Symbol("[[webidl.brand]]")]: Symbol("[[webidl.brand]]"),
    [Symbol("url object")]: null,
    [Symbol(list)]: [
      [
        "list",
        "PLmoRDY8IgE2Okxy4WWdP95RHXOTGzJfQs",
      ],
    ],
  },
  pathType: "list",
}
`;

snapshot[`youtube links > is 3`] = `
{
  params: URLSearchParams {
    [Symbol("[[webidl.brand]]")]: Symbol("[[webidl.brand]]"),
    [Symbol("url object")]: URL {
      hash: "",
      host: "www.youtube.com",
      hostname: "www.youtube.com",
      href: "https://www.youtube.com/watch?v=57rdbK4vmKE&list=PLmoRDY8IgE2Okxy4WWdP95RHXOTGzJfQs",
      origin: "https://www.youtube.com",
      password: "",
      pathname: "/watch",
      port: "",
      protocol: "https:",
      search: "?v=57rdbK4vmKE&list=PLmoRDY8IgE2Okxy4WWdP95RHXOTGzJfQs",
      username: "",
    },
    [Symbol(list)]: [
      [
        "v",
        "57rdbK4vmKE",
      ],
      [
        "list",
        "PLmoRDY8IgE2Okxy4WWdP95RHXOTGzJfQs",
      ],
    ],
  },
  pathType: "com",
  videoId: "57rdbK4vmKE",
}
`;

snapshot[`youtube links > is 4`] = `
{
  params: URLSearchParams {
    [Symbol("[[webidl.brand]]")]: Symbol("[[webidl.brand]]"),
    [Symbol("url object")]: URL {
      hash: "",
      host: "music.youtube.com",
      hostname: "music.youtube.com",
      href: "https://music.youtube.com/watch?v=nj1cre2e6t0",
      origin: "https://music.youtube.com",
      password: "",
      pathname: "/watch",
      port: "",
      protocol: "https:",
      search: "?v=nj1cre2e6t0",
      username: "",
    },
    [Symbol(list)]: [
      [
        "v",
        "nj1cre2e6t0",
      ],
    ],
  },
  pathType: "com",
  videoId: "nj1cre2e6t0",
}
`;

snapshot[`youtube links > is 5`] = `
{
  params: URLSearchParams {
    [Symbol("[[webidl.brand]]")]: Symbol("[[webidl.brand]]"),
    [Symbol("url object")]: null,
    [Symbol(list)]: [],
  },
  pathType: "dotbe",
  videoId: "nj1cre2e6t0",
}
`;

snapshot[`youtube links > is not 1`] = `undefined`;

snapshot[`youtube links > is not 2`] = `undefined`;

snapshot[`youtube links > is not 3`] = `undefined`;

snapshot[`youtube links > is not 4`] = `undefined`;
